### DISCUSSION TITLE
Massive Rigidbody3D velocity reset from ground friction after airbourne?

### DISCUSSION THREAD

**Discussion by Mangonels** [godot-jolt/godot-jolt#915] [Questions]
I have an issue with my player controlled main character when falling on a StaticBody3D.

The main character is formed by two RigidBody3D's, (but I must be clear on the fact that the issue would still ocur if it were a single rigidbody), one for the friction with the ground (at the character's feet), which has a physics material with 0.7 friction (It has a sphere collision as child that's constantly colliding with the ground), and one for the rest of the body (with a capsule collision), which has 0.0 friction (so it doesn't get stuck to walls and other rigidbodies slide off of it).

https://youtu.be/CMVlzhKRiUI

In this video you will notice the rigidbody practically loses all of it's velocity (sometimes it actually does go to 0 velocity) as soon as it lands on the ground diagonally from a jump. 

It then regains this velocity naturally due to continuously applied force while grounded, each physics process, but the result is a very visual velocity reduction on the main character. I'm using phantom camera by the way, so it's much more visible since the camera and the visuals are smooth. On a regular godot camera the velocity "stoppage" would still be noticeable by carefully staring at the background.

Coming from Unity's 3D physics, this is something that didn't happen with the same setup over there.
Additionally, I don't recall this issue with godot's default physics. I remember I started experiencing this as soon as I installed and toggled on Jolt.

So I'm suspecting how the friction system is coded internally in Jolt has something to do with this, but I may be wrong.

Another option is that my code is faulty. It's quite likely I could design it better with your advice, but at the end of the day after all my calculations, the only thing that's going on is that I'm adding uninterrupted force (as you can see in the logs which may vary slightly from airbourne to grounded, but is always substantial enough), which further points to this issue having something to do with friction.

I also tried tweaking the value of friction for the feet, but too low will make the character too slippery, and you can also encounter this velocity reset with some surprisingly low (below the shown 0.7 in the video) feet friction values. In any case I still tried setting the friction to 0.0, and the velocity values do indeed seem to be reduced on landing, but not as harsh visually.

**Comment by mihe** [comment]
Any chance I could ask you to provide a minimal reproduction project of some sort? I can't say from just the video alone what the problem might be, even if it clearly showcases the issue you're describing.

The more you can reduce the script/scene/project down the better. That includes any modified project settings and whatnot as well.

> Additionally, I don't recall this issue with godot's default physics. I remember I started experiencing this as soon as I installed and toggled on Jolt.

Are you able to switch over to Godot Physics to see if the same thing happens there?

**Reply by Mangonels** [reply]
> Any chance I could ask you to provide a minimal reproduction project of some sort? I can't say from just the video alone what the problem might be, even if it clearly showcases the issue you're describing.
> 
> The more you can reduce the script/scene/project down the better. That includes any modified project settings and whatnot as well.
> 
> > Additionally, I don't recall this issue with godot's default physics. I remember I started experiencing this as soon as I installed and toggled on Jolt.
> 
> Are you able to switch over to Godot Physics to see if the same thing happens there?

I made a "minimum reproduction project" from scratch, but I added some movement code from my original project.
I think the bulkiest parts right now might be the movement code, and the fact that I installed Jolt and PhantomCamera plugin.

https://github.com/Mangonels/character_movement

Switching to godot physics makes the issue barely noticeable but according to the logs there's also a certain degree of velocity reduction. Definetly not as pronounced as with Jolt though.

**Reply by mihe** [reply]
Thank you. I'll take a look at this some time in the next few days.

**Reply by Mangonels** [reply]
> Thank you. I'll take a look at this some time in the next few days.

Hi, I couldn't figure out what's going on or mitigate the issue by changing settings values, will this still be looked at?

**Reply by mihe** [reply]
Yes, sorry, this is on my to-do list (among a bunch of other things) but I haven't gotten around to looking at your repro yet.

Just out of curiosity (without having looked at this yet) does the discrepancy between Jolt and Godot Physics go away if you bump up the `physics/jolt_3d/solver/position_correction` project setting to 100%?

**Reply by Mangonels** [reply]
> Yes, sorry, this is on my to-do list (among a bunch of other things) but I haven't gotten around to looking at your repro yet.
> 
> Just out of curiosity (without having looked at this yet) does the discrepancy between Jolt and Godot Physics go away if you bump up the `physics/jolt_3d/solver/position_correction` project setting to 100%?

Ok! Thanks for letting me know, I'm actually not in a hurry, knowing this is in your todo list is good enough for now! :)
I also wanted to make it clear that I'm alert for any tips or solutions for this issue.

Yes I did change `physics/jolt_3d/solver/position_correction`, I tried 25%, 50% (my standard value) and 100%, no luck.

It's not really a discrepancy. Godot does also apply a certain amount of velocity reduction on landing, however this is MUCH more pronounced with Jolt. (in fact, it's one of the first things I noticed when switching to Jolt, that rigidbodies seem to stick to the ground on landing much easier, it looks to be derived from friction calculation).

**Comment by Mangonels** [comment]
Hello, I updated the reproduction project to Godot 4.4 and uninstalled godot-jolt extension + set the integrated jolt physics. 
I can still reproduce this issue: https://github.com/Mangonels/Godot-RigidBody3D-character-movement

I must also add that I recently started a godot-cpp project where I constructed some rigidbodies directly through PhysicsServer and I couldn't quite notice this landing deceleration there (but I can't test/debug the velocity there right now because I also updated that project to 4.4 and it's setup is broken for now).

I may open an issue on the godot github if you find it fine, even if it turns out direct PhysicsServer through godot-cpp actually does work fine vs my gdscript rigidbody character controller implementation, because I really don't know what's causing it.

**Reply by mihe** [reply]
> I may open an issue on the godot github if you find it fine

Please do! Sorry, I had completely forgotten about this issue.

Realistically I'll be the one to take a look at the issue anyway, but who knows, maybe someone else finds the time to do so.

**Reply by mihe** [reply]
Thank you for [the upstream issue](https://github.com/godotengine/godot/issues/104380). I'll go ahead and close this discussion/issue as duplicate in that case.
